File: src/types/library.ts
export interface Track {
  id: string;
  title?: string;
  artist?: string;
  album?: string;
  genre?: string;
  bpm?: number;
  key?: string;
  rating?: number;
  comments?: string;
  location?: string;
}

export interface Playlist {
  id: string;
  name: string;
  kind: "folder" | "playlist";
  trackIds: string[];
  children: Playlist[];
}

export interface LibraryMetadata {
  rekordboxVersion?: string;
  productName?: string;
  productVersion?: string;
}

export interface Library {
  tracks: Track[];
  playlists: Playlist[];
  metadata: LibraryMetadata;
}

export interface LoadedLibrary {
  sourcePath: string;
  library: Library;
}

export interface SaveLibraryRequest {
  destinationPath: string;
  sourcePath?: string;
  library: Library;
  createBackup: boolean;
}

export interface SaveLibraryResult {
  destinationPath: string;
  backupPath?: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
}

export interface BulkEditPatch {
  artist?: string;
  album?: string;
  genre?: string;
  key?: string;
  rating?: number;
  comments?: string;
}

File: apps/tauri/src-tauri/src/models/library.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Track {
    pub id: String,
    pub title: Option<String>,
    pub artist: Option<String>,
    pub album: Option<String>,
    pub genre: Option<String>,
    pub bpm: Option<f32>,
    pub key: Option<String>,
    pub rating: Option<u8>,
    pub comments: Option<String>,
    pub location: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Playlist {
    pub id: String,
    pub name: String,
    pub kind: PlaylistKind,
    pub track_ids: Vec<String>,
    pub children: Vec<Playlist>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PlaylistKind {
    Folder,
    Playlist,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LibraryMetadata {
    pub rekordbox_version: Option<String>,
    pub product_name: Option<String>,
    pub product_version: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Library {
    pub tracks: Vec<Track>,
    pub playlists: Vec<Playlist>,
    pub metadata: LibraryMetadata,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LoadedLibrary {
    pub source_path: String,
    pub library: Library,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveLibraryRequest {
    pub destination_path: String,
    pub source_path: Option<String>,
    pub library: Library,
    pub create_backup: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveLibraryResult {
    pub destination_path: String,
    pub backup_path: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ValidationResult {
    pub valid: bool,
    pub errors: Vec<String>,
}
